22_Introduction 14th October 2019

Basic 10x RNA-seq analysis using R
Packages: Seurat/scran/Monocole/Pagoda2/Liger(Macosko lab)

Chromium 3' Single Cell Workflow (Bielas et al Nature Communication 2017)
10x provides Cell Ranger - summary HTML
Cell Ranger Output - barcodes.tsv (cell containing barcodes), genes or features.tsv, matrix.mtx (total genes and total cells)
Other tools e.g Alevin (Salmon) for alignment free single cell RNA-Seq (alternative to Cell Ranger)

Basic pipeline for 10x single cell analysis:
- reading in 3 matrices and transforming data into matrix ("gene" as a row, "barcode" or "cell" as column)

- data filtering on user defined QC metrics (genes that are expressed in >=3 cells (~1% data), highly variable genes) 
  (cells min >=200 number of detected genes per cell)
  (genes from mitochondrial or ribosomal genes? Should be low proportion)
  QC metrics: identify low quality cells - Ilicic et al Genome Biology 2016

- data normalisation (removing cell specific biases) - Seurat global scaling normalisation, Scater/scran small datasets

- finding highly variable genes and scale data

- remove confounding factors (eg batch, gender, cell size, cell cycle). Remove cell cycle phases especially for immune cell data,
  otherwise data will cluster by phases rather than actual sample. Don't always have to regress out, dependent on analysis type.

- dimension reduction (PCA plot - ICA, CCA) - p values to see which PC is significant

- clustering

- tSNE

- find differentially expressed genes between clusters

R code below from session - convert to .Rmd later.

# Basic single-cell RNA-seq analysis using Seurat
# Author: Devika Agarwal
# Contact : devika.agarwal@psych.ox.ac.uk 
#############################################
# R codes are mainly from https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html 
#############################################
library(Seurat)
library(tidyverse)
library(clustree)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "single_cell/filtered_gene_bc_matrices/hg19") 
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
############################################
# Lets examine a few genes in the first thirty cells
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]

##################################################
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
#We calculate mitochondrial QC metrics with the PercentageFeatureSet function, 
#which calculates the percentage of counts originating from a set of features
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

# Show QC metrics for the first 5 cells
head(pbmc@meta.data, 5)

############################################
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))
#################Filtering##################
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

############################################
#Normalized values are stored in  pbmc[["RNA"]]@data
pbmc <- NormalizeData(object = pbmc, normalization.method = "LogNormalize", 
                      scale.factor = 10000)
############################################
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))

#length(x = pbmc@var.genes)
############################################
#The results of this are stored in pbmc[["RNA"]]@scale.data
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes,vars.to.regress = "percent.mt")
# can include percent.ribo, batch effect , cell cycle and so on

##########Perform linear dimensional reduction############################
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")

DimPlot(pbmc, reduction = "pca")
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)

#pbmc <- JackStraw(pbmc, num.replicate = 100)
#pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
#JackStrawPlot(pbmc, dims = 1:15)

ElbowPlot(pbmc)
#################Clustering Analysis###############
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)

#PrintFindClustersParams(object = pbmc)
#As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.
pbmc <- RunTSNE(pbmc, dims = 1:10)

p1 <-TSNEPlot(pbmc)

#######################################
#Run umap
#######################################
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
pbmc <- RunUMAP(pbmc, dims = 1:10, label=T)

# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
p2 <- DimPlot(pbmc, reduction = "umap")

CombinePlots(plots = list(p1,p2) , ncol =2)
##library(cowplot)
#p1 <- DimPlot(object = pbmc, reduction = "tsne", no.legend = TRUE, do.return = TRUE, 
 #             vector.friendly = TRUE, pt.size = 0.1) + ggtitle("tSNE") + theme(plot.title = element_text(hjust = 0.5))
#p2 <- DimPlot(object = pbmc, reduction = "umap", no.legend = TRUE, do.return = TRUE, 
#              vector.friendly = TRUE, pt.size = 0.1) + ggtitle("UMAP") + theme(plot.title = element_text(hjust = 0.5))
#plot_grid(p1, p2)



# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
#########Plot genes######################
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))

# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
#########################################
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", 
                               "CD8A"))

########################################
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()

##################################################
# Assigning cell type identities to cluster
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", 
                     "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()

#################################
saveRDS(pbmc, file = "pbmc3k_final.rds")

