Stages of data processing and appropriate downstream applications
Measured:
  Raw - statistical testing (differential gene expression)
  Normalised (+ log transformed)
Corrected:
  Corrected (technical) - visual comparison of data
  Corrected (bioogical) - pre-processing for trajectory inference
Reduced:
  Featured selected - visualisation, trajectory inference
  Dimensionality reduced - visualisation, clustering, trajectory inference

Downstream analyses after QC and normalisation:
Cell level - cluster analysis, trajectory analysis, compositional analysis, metastable states,
  cluster annotation, gene expression dynamics
Gene level - differential expression analysis, gene set analysis, gene regulatory networks

Clustering - https://cran.r-project.org/web/packages/clustree/vignettes/clustree.html

# Files and data for exercises saved in OBDS Training folder on Desktop (on laptop)

# Exercise 1: use clustertree to evaluate cluster resolutions for package example data or use the PBMC 3k data from Seurat
# (clustering already done in Seurat object)

# can view summary of data and resolutions available
head(pbmc[[]])

# by default only have 0.5 resolution (saved under metadata) - use for loop to set more resolutions of 0.1 step
for (ii in seq(0.1, 1, 0.1)){
  pbmc = FindClusters(object = pbmc, resolution = ii)
}

# this will now be in metadata - can cluster using clustree at different resolutions
# use this to select ideal resolution
clustree(pbmc, prefix="RNA_snn_res.")

# Numbers in cluster tree relate to new.cluster.ids

# Exercise 2: Sub-cluster annotated clusters
# probably look at Alistair's code?
# https://satijalab.org/seurat/essential_commands.html - object interaction tab

library(Seurat)
library(SingleCellExperiment)
library(tidyverse)

# read file in
baron_human <- readRDS(file="baron-human.rds")

# create Seurat object
baron.seurat <- as.Seruat(baron_human)

# normalise data
baron.seurat <- NormalizeData(object=baron.seurat, normalization.method="LogNormalize", scale.factor=10000)

# find variable features
baron.seurat <- FindVariablesFeatures(baron.seurat)

# centering and scaling data matrix
baron.seurat <- ScaleData(baron.seurat)

# get cell identity classes (dataset already annotated)
Idents(baron.seurat) <- "cell_type1"
levels(x=baron.seurat)

# subset Seurat object based on identity class
acinar <- subset(x=baron.seurat, idents="acinar")
acinar <- FindVariableFeatures(acinar, selection.method="vst", nfeatures=2000)
top10 <- head(VariableFeatures(acinar), 10)

plot1 <- VariableFeaturePlot(acinar)
plot2 <- LabelPoints(plot=1, points=top10, repel=T)

CombinePlots(plots=list(plot1, plot2))

acinar <- RunTSNE(acinar, dims=1:10)
p1 <- TSNEPlot(acinar)

acinar <- RunUMAP(acinar, dims=1:10, label=T)
p2 <- DimPlot(acinar, reduction="umap")

CombinePlots(plots=list(p1, p2), ncol=2)


# Exercise 3: Mapping of human onto mouse data (Alistair's code)
# Alignment from https://satijalab.org/seurat/v3.1/integration.html
# Imports
```R
library(tidyverse)
library(Seurat)
library(biomaRt)
```

# Functions
# Load Data
```R
sce_mouse = readRDS('baron-mouse.rds')
sce_human = readRDS('baron-human.rds')
```
# Get gene symbols from biomart
```R
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
 
gene_symbol_conversion = getLDS(attributes = c("hgnc_symbol"), 
                         filters = "hgnc_symbol", 
                         values = rownames(sce_human),
                         mart = human, 
                         attributesL = c("mgi_symbol"), 
                         martL = mouse, 
                         uniqueRows=T)
```

```R
head(gene_symbol_conversion)
```
# Take the counts table from the singleCellExperiment object
```R
sce_mouse_count <- counts(sce_mouse)
```

# Convert counts table to dataframe and merge with the conversion table generated from biomart
Here I'm merging on the mouse symbols

```R
df_sce_mouse_count = rownames_to_column(as.data.frame(sce_mouse_count), var='symbol')
sce_mouse_count = merge(df_sce_mouse_count, gene_symbol_conversion, by.x='symbol', by.y='MGI.symbol')
```

# Gene symbols have duplicates therefore need to aggregate the values for the duplicates together

To do this: 
* Drop the first column (contains the original rownames)
* Group by the human symbol
* Calculate the mean for all of the columns that are aggregated in the groupby
(This function takes a long time to run)

```R
sce_mouse_count = sce_mouse_count[,-1]
sce_mouse_count = sce_mouse_count %>%
                  group_by(HGNC.symbol) %>%
                  summarise_all(mean) 
```

# Convert the human gene symbol to rownames
```R
sce_mouse_count = column_to_rownames(sce_mouse_count, var = 'HGNC.symbol')
```

# Generate a new SCE object - converted mouse symbols to human symbols
```R
sce_mouse_new = SingleCellExperiment(assays = list(counts = as.matrix(sce_mouse_count)),
                                     colData = colData(sce_mouse))
```
```R
sce_mouse_new
```
# Calculating log counts
libsizes <- colSums(sce_mouse_count)
size.factors <- libsizes/mean(libsizes)
logcounts(sce_mouse_new) <- log2(t(t(sce_mouse_count)/size.factors) + 1)
assayNames(sce_mouse_new)

# Making Seurat object and combining to list

so_mouse = as.Seurat(sce_mouse_new)
so_human = as.Seurat(sce_human)

so_human@meta.data$organism = "human"
so_mouse@meta.data$organism = "mouse"

so_list = c(so_mouse, so_human)

for (ii in 1:length(so_list)){
    so_list[[ii]] <- NormalizeData(so_list[[ii]], verbose = FALSE)
    so_list[[ii]] <- FindVariableFeatures(so_list[[ii]], selection.method = "vst", 
                                         nfeatures = 2000, verbose = FALSE)
}

# Integrate data
so_anchors <- FindIntegrationAnchors(object.list = so_list, dims = 1:30, )
so_integrated <- IntegrateData(anchorset = so_anchors, dims = 1:30)

# Process integrated data
DefaultAssay(so_integrated) <- "integrated"
so_integrated <- ScaleData(so_integrated, verbose = FALSE)
so_integrated <- RunPCA(so_integrated, npcs = 30, verbose = FALSE)
so_integrated <- RunUMAP(so_integrated, reduction = "pca", dims = 1:30)

# plot combined data
Idents(so_integrated) = 'cell_type1'
p1 = DimPlot(so_integrated, reduction = 'umap', group.by = "organism", 
             label.size=4, repel= TRUE)
p1

p2 = DimPlot(so_integrated, reduction = 'umap', group.by = "cell_type1", 
             label = TRUE, label.size=4, repel= TRUE)
p2
