# Network Analysis 17th October 2019

## Biological Networks - basic components
  Nodes (circles): gene/protein/metabolite/ontology
  Edge: line connecting 2 nodes
  Activation/inhibition (direct/indirect)
## Biological Networks - basic features/network topology
  Degree: number of connections of a node
  Shortest paths or distance
  Hubs: high degree nodes
  Scale free networks: small number of neighbours, small number of hubs. Biological networks unlikely to be scale free?
  A single gene knocked out in a network is unlikely to lead to a lethal effect due to redundancy; however knocking out a hub
  gene could lead to lethality eg P53 in cancer is a hub.
  Clusters/subcommunities: group of nodes more internally connected than they are with rest of network
  Centralities: estimate of how important node/edge is for connectivity
  Size/density of network; motifs/cliques/clusters/sub-networks
## Types/sources of biological networks:
  DNA-protein (transcriptional regulatory networks, methylation)
  RNA-RNA (miRNA regulatory)
  RNA-Protein (splicing regulatory)
  Gene-Gene (co-expression)/Protein-Protein (coexpression, colocalisation, Gene Ontology etc)
 
 What's the point of building/analysing biological networks?
  Functional features, meta-analysis.

Ingenuity Pathway Analysis - not free, but they do a lot of literature search to back up pathways

## WGCNA: weighted gene correlation network analysis
  construct gene co-expression network
  identify modules
  relate modules to external information
  study module relationships
  find key drivers in interesting molecules
  https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials
  Ran code in R Studio in X2Go. Will export for comments.
  
# Remember in R Studio to clear cache before Knit to HTML!
